vernac/vernacentries.ml

```diff
@@ -506,7 +506,174 @@ let sort_universes g =
+      else
+        let visited = UMap.add source (k <> Univ.Lt) visited in
+        let visited, next_todo =
+          UMap.fold (fun u (ref,k') (visited,next_todo) ->
+              if k = Univ.Eq && k' = Univ.Lt then
+                (* no point searching for a loop involving [u]  *)
+                (UMap.add u true visited, next_todo)
+              else
+                let next_k = if k = Univ.Lt && k' = Univ.Lt then Univ.Le
+                  else k
+                in
+                let revpath = (k',u,ref) :: revpath in
+                if Univ.Level.equal u target && next_k <> Univ.Lt
+                then raise (Found revpath)
+                else (visited, (u, next_k, revpath) :: next_todo))
+            (Option.default UMap.empty (UMap.find_opt source src.edges))
+            (visited,next_todo)
+        in
+        loop visited todo next_todo
+  in
+  try loop UMap.empty [source,k,[]] []; None
+  with Found l -> Some (List.rev l)
+
+let search src (u,k,v) =
+  let path = search src ~source:u k ~target:v in
+  match path with
+  | None -> None
+  | Some path ->
+    if k = Univ.Eq && not (List.for_all (fun (k',_,_) -> k' = Univ.Eq) path) then
+      let path' = search src ~source:v k ~target:u in
+      begin match path' with
+      | None -> None
+      | Some path' -> Some (path @ path')
+      end
+    else Some path
+
+let find_source (u,k,v as cst) src =
+  if Univ.Level.is_set u && k = Univ.Lt then []
+  else Option.default [] (search src cst)
+
+let pr_constraint_source = function
+  | GlobRef ref -> pr_global ref
+  | Library dp -> str "library " ++ pr_qualid (Nametab.shortest_qualid_of_module (MPfile dp))
+
+let pr_source_path prl u src =
+  if CList.is_empty src then mt()
+  else
+    let pr_rel = function
+      | Univ.Eq -> str"=" | Lt -> str"<" | Le -> str"<="
+    in
+    let pr_one (k,v,ref) =
+      spc() ++
+      h (pr_rel k ++ surround (str "from " ++ pr_constraint_source ref) ++
+         spc() ++ prl v)
+    in
+    spc() ++ surround (str"because" ++ spc() ++ prl u ++ prlist_with_sep mt pr_one src)
+
+let pr_pmap sep pr map =
+  let cmp (u,_) (v,_) = Univ.Level.compare u v in
+  Pp.prlist_with_sep sep pr (List.sort cmp (Univ.Level.Map.bindings map))
+
+let pr_arc srcs prl = let open Pp in
+  function
+  | u, UGraph.Node ltle ->
+    if Univ.Level.Map.is_empty ltle then mt ()
+    else
+      prl u ++ str " " ++
+      v 0
+        (pr_pmap spc (fun (v, strict) ->
+             let k = if strict then Univ.Lt else Univ.Le in
+             let src = find_source (u,k,v) srcs in
+             hov 2 ((if strict then str "< " else str "<= ") ++ prl v ++ pr_source_path prl u src))
+            ltle) ++
+      fnl ()
+  | u, UGraph.Alias v ->
+    let src = find_source (u,Eq,v) srcs in
+    prl u  ++ str " = " ++ prl v ++ pr_source_path prl u src ++ fnl ()
+
+let pr_universes srcs prl g = pr_pmap Pp.mt (pr_arc srcs prl) g
+
+let print_universes { sort; subgraph; with_sources; file; } =
```

Comment:
Wouldn't it be worth having a new vernac/comPrintUniverse.ml file or the like rather than adding that in this already large file?